### PR TITLE
Guard against OIDC Provider configuration being stale

### PR DESCRIPTION
**Pull request by segiddins**
Fixes TOB-RGM-14

### CODE REVIEW

File: app/controllers/concerns/jwt_validation.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    %w[iss jti].each do |claim|
+      raise InvalidJWT, "Missing/invalid #{claim}" unless @jwt[claim].is_a?(String)
+    end
+  end
+
+  def validate_provider
+    raise UnsupportedIssuer, "Provider is missing jwks" if @provider.jwks.blank?
+    # TODO: delete &. after all providers have updated their configuration
+    return unless @provider.configuration_updated_at&.before?(1.day.ago)
+    raise UnsupportedIssuer, "Configuration last updated too long ago: #{@provider.configuration_updated_at}"
```
Comment (martinemde) [inline app/controllers/concerns/jwt_validation.rb:43]: When this error happens there's nothing the user can do about it, no?

Should we indicate this with a 503, in effect saying "it's not you it's me".

File: app/controllers/concerns/jwt_validation.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    render_bad_request(e)
+  end
+
+  def validate_jwt_format
+    %w[nbf iat exp].each do |claim|
+      raise InvalidJWT, "Missing/invalid #{claim}" unless @jwt[claim].is_a?(Integer)
+    end
+    %w[iss jti].each do |claim|
+      raise InvalidJWT, "Missing/invalid #{claim}" unless @jwt[claim].is_a?(String)
+    end
```
Comment (martinemde) [inline app/controllers/concerns/jwt_validation.rb:36]: Nit: could extract to constant `{iss: String, ...}` which would allocate once and be a little simpler.

File: app/controllers/concerns/jwt_validation.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      with: :render_not_found
+    )
+  end
+
+  def jwt_key_or_secret
+    raise NotImplementedError
+  end
+
+  def decode_jwt
+    @jwt = JSON::JWT.decode_compact_serialized(params.permit(:jwt).require(:jwt), jwt_key_or_secret)
```
Comment (martinemde) [inline app/controllers/concerns/jwt_validation.rb:24]: Is `:skip_verification` handled by the JWT lib?

**Review by martinemde — APPROVED** [review]
Mostly just style nitpicks. I like the module approach. Much cleaner.

File: app/controllers/concerns/jwt_validation.rb (hunk 3 above)
Comment (segiddins) [reply app/controllers/concerns/jwt_validation.rb:24]: yes

**Comment by yob** [comment]
I believe this has broken using API Key Roles with Buildkite's OIDC provider. I've been trying to publish a gem today, and the assume_role API call is returning an error:

```
{"error":"Missing/invalid jti"}
```

That's not particularly surprising, since Buildkite's OIDC tokens don't include a `jti` claim. Is that a new requirement for the `assume_role` endpoint, or is the new requirement there an accidental side effect of refactoring to tighten the security for the newer trusted publishers behaviour?

**Comment by yob** [comment]
@segiddins thanks! 

I'm looking into whether we can add a `jti` claim to our tokens. We probably can, it seems pretty straight forward. I'll wait to see if that happens, then open a new issue with a sanitised Buildkite OIDC token than can be added to a regression test :heart:

**Comment by yob** [comment]
Buildkite OIDC tokens now have a `jti` claim, and I've confirmed they work correctly with the API Key Roles feature. I'll open a separate issue with a sample token that can be used to help avoid future regressions ❤️

**Pull request merged**